### ISSUE TITLE
RSYNC backup to remote storage not working

### ISSUE THREAD

**Issue by andyaz** [FalconChristmas/fpp#1529]
While attempting to use the remote RSYNC file copy backup function from a remote FPP instance to the USB SSD attached to the Master FPP (mounts as sdb1) the process starts but never finishes or transfers any files after the file list sent. 

I have attached the logs from both the remote and the master FPP. 
rsync is enabled in setup on both FPP instances. 
All FPP Pis have been freshly rebooted.
Local file copy backups on the master to the local USB SSD work properly.
Hardware: Pi 3b+
All on same network segment 192.168.128.0/24 with no firewalls or routers in the middle.

Versions info:
Remote FPP:
FPP Version: | 7.x-master-138-g85ee4d65
Platform: | Raspberry Pi (Pi 3 Model B)
FPP OS Build: | v2023-06
OS Version: | Raspbian GNU/Linux 11 (bullseye)
Hardware Serial Number: | 000000001478da20
Kernel Version: | 6.1.21-v7+
System Boot Time: | 2023-06-27 22:15:22
fppd Uptime: | 45:46
Local Git Version: | 85ee4d6 ChangeLog
Remote Git Version: | 85ee4d6

Master FPP:
Version Info:
FPP Version: | 7.x-master-138-g85ee4d65
Platform: | Raspberry Pi (Pi 3 Model B)
FPP OS Build: | v2023-04
OS Version: | Raspbian GNU/Linux 11 (bullseye)
Hardware Serial Number: | 0000000090a476ea
Kernel Version: | 6.1.21-v7+
System Boot Time: | 2023-06-27 22:15:29
fppd Uptime: | 54:52
Local Git Version: | 85ee4d6 ChangeLog
Remote Git Version: | 85ee4d6

[FPP_Logs_20230628.0559.zip](https://github.com/FalconChristmas/fpp/files/11890696/FPP_Logs_20230628.0559.zip)
[FPP_Logs_20230628.0429.zip](https://github.com/FalconChristmas/fpp/files/11890698/FPP_Logs_20230628.0429.zip)
![image](https://github.com/FalconChristmas/fpp/assets/19541773/112f16f6-3075-4f0d-97e9-95224455d071)
![image](https://github.com/FalconChristmas/fpp/assets/19541773/668b8542-0cd2-4d0f-9d2a-babe32b7d8e0)

**Comment by andyaz** [comment]
Interesting thing I learned playing with it today. If I select everything except for Virtual EEPROM it successfully backs up. Since this is not a beagle there of course would be no virtual EEPROM... so something related to that is causing the hangup. Maybe some sort of check in the process that if there is no EE and it is selected it would just exclude that and continue on with the process?

**Comment by andyaz** [comment]
More information - I have tried it from 3 other FPP instances and I am finding even with only sequences and configuration checked (which I know those Pis have on them) it is mounting but not successfully backing up so there is still something fishy going on... I also have to reboot the master and the failed remote when things go wrong in order to get the master sdb1 to show up in the remote list in any other instance.

**Comment by jaredb7** [comment]
> I also have to reboot the master and the failed remote when things go wrong in order to get the master sdb1 to show up in the remote list in any other instance.

Yeah unfortunately this happens because the drive is still mounted for the remote USB file copy operation on the destination/master, in normal operation it gets unmounted at the end the process. Though in situations like this where it doesn't get a chance to do that it causes issues.  
I'll put in a unmount when you close the dialogue and also when you visit the FPP Backups page, that should then fix any issues where the operation is interrupted for some reason.

Also it shouldn't matter what backup items are selected as to whether it succeeds or not. In the example of Virtual EEPROM's it just continues if there's nothing. May have just been luck it worked that time it was unselected :)

I tried to see if I could reproduce this but everything produces some sort of output instead of just nothing after the mount
- If the drive doesn't mount - rsync spits out tonnes of errors about not being able to write to the expected location.
- Same with disconnecting/dropping the USB storage after it's mounted - rsync outputs different but more or less the same I\O errors. 

Digging into the logs a bit further I'm if the voltage condition is doing something weird.., I'm not sure if it is though, but should try rule it out. 

Could you try to see if you can use put another USB stick on the master and do a backup onto that? Just to try something basic?


**Log Snippets**
(not a an actual error though)
> 
[Tue Jun 27 22:18:37.908321 2023] [proxy_fcgi:error] [pid 1530:tid 1728050176] [client 192.168.128.54:49638] AH01071: Got error 'PHP message: MountDevice: ( {"Status":"OK","Message":"Device (sdb1) mounted at (\\/mnt\\/remote_filecopy)","MountLocation":"\\/mnt\\/remote_filecopy"} )'

**syslog**
> Jun 27 22:17:14 FPP-MASTER kernel: [  105.448210] hwmon hwmon1: Voltage normalised
> Jun 27 22:18:01 FPP-MASTER CRON[3360]: (fpp) CMD (/usr/bin/php /home/fpp/media/plugins/fpp-after-hours/cronMonitorStream.php >/dev/null 2>&1)
> **Jun 27 22:18:37 FPP-MASTER kernel: [  188.884952] exFAT-fs (sdb1): Volume was not properly unmounted. Some data may be corrupt. Please run fsck.**
> Jun 27 22:18:39 FPP-MASTER kernel: [  190.727878] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:18:48 FPP-MASTER kernel: [  199.047350] hwmon hwmon1: Voltage normalised

**syslog**
And quite a few more occurrences afterwards, but these messages are appearing all through the syslog
> Jun 27 22:19:00 FPP-MASTER kernel: [  211.529257] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:19:01 FPP-MASTER CRON[3470]: (fpp) CMD (/usr/bin/php /home/fpp/media/plugins/fpp-after-hours/cronMonitorStream.php >/dev/null 2>&1)
> Jun 27 22:19:12 FPP-MASTER kernel: [  224.013180] hwmon hwmon1: Voltage normalised
> Jun 27 22:19:15 FPP-MASTER kernel: [  226.093732] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:19:19 FPP-MASTER kernel: [  230.254615] hwmon hwmon1: Voltage normalised
> Jun 27 22:19:35 FPP-MASTER kernel: [  246.897474] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:19:40 FPP-MASTER kernel: [  251.058173] hwmon hwmon1: Voltage normalised
> Jun 27 22:19:42 FPP-MASTER kernel: [  253.138176] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:20:00 FPP-MASTER kernel: [  271.859664] hwmon hwmon1: Voltage normalised
> Jun 27 22:20:01 FPP-MASTER CRON[4950]: (fpp) CMD (/usr/bin/php /home/fpp/media/plugins/fpp-after-hours/cronMonitorStream.php >/dev/null 2>&1)
> Jun 27 22:20:02 FPP-MASTER kernel: [  273.939783] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:20:11 FPP-MASTER kernel: [  282.260115] hwmon hwmon1: Voltage normalised
> Jun 27 22:20:13 FPP-MASTER kernel: [  284.340191] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:20:17 FPP-MASTER kernel: [  288.500263] hwmon hwmon1: Voltage normalised
> Jun 27 22:20:19 FPP-MASTER kernel: [  290.580373] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:20:23 FPP-MASTER kernel: [  294.740430] hwmon hwmon1: Voltage normalised
> Jun 27 22:20:27 FPP-MASTER kernel: [  298.900726] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:20:48 FPP-MASTER kernel: [  319.700854] hwmon hwmon1: Voltage normalised
> Jun 27 22:21:01 FPP-MASTER kernel: [  332.181008] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:21:01 FPP-MASTER CRON[5418]: (fpp) CMD (/usr/bin/php /home/fpp/media/plugins/fpp-after-hours/cronMonitorStream.php >/dev/null 2>&1)
> Jun 27 22:21:05 FPP-MASTER kernel: [  336.341004] hwmon hwmon1: Voltage normalised
> Jun 27 22:21:07 FPP-MASTER kernel: [  338.421047] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:21:19 FPP-MASTER kernel: [  350.901045] hwmon hwmon1: Voltage normalised
> Jun 27 22:21:24 FPP-MASTER kernel: [  355.061108] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:21:28 FPP-MASTER kernel: [  359.221104] hwmon hwmon1: Voltage normalised
> Jun 27 22:21:44 FPP-MASTER kernel: [  375.861194] hwmon hwmon1: Undervoltage detected!
> Jun 27 22:21:51 FPP-MASTER kernel: [  382.101191] hwmon hwmon1: Voltage normalised

**Comment by andyaz** [comment]
I ran FSCK on the drive and fixed the dirty bit with no other errors detected.
I could see power being a possible issue with the original Canakit 3b+ supply (2.5A I think) trying to run the pi and the SSD. I ordered some 4amp power supplies that should be here later this week. 
I also have managed to get all but 1 pi to back up successfully now... it is very unpredictable it seems to make sense that power might be the answer... I will try again once I get them here and let you know the results.

**Comment by jaredb7** [comment]
Thanks for trying to rule those things out and running FSCK to get everything happy... I was fairly sure that wasn't causing much issue unless the file system was corrupt (I'm sure we'd be seeing other more specific errors if it was).

I was able to reproduce the issue with a by having the FPP OS Upgrade file downloaded (just luck I had it here from reimaging and upgrading through to 7.x Beta), the copy is actually going along perfectly fine but there's no output coming back to the browser up. The first rsync copy is Configuration and the uploads directory gets copied across at this stage, so any large files here would appear as no progress or output after the initial mounting.

I can imagine if you have a few gigabytes worth of files to copy across that there would be no output for a some time across various folders (videos, sequences etc) and thus looking like it hung

I'll investigate further.
![image](https://github.com/FalconChristmas/fpp/assets/799998/51cf93f0-2bb6-4983-af55-bd9eadba434e)

![image](https://github.com/FalconChristmas/fpp/assets/799998/14b37b68-6854-48f2-bf9e-1922b0d2a639)

**Comment by andyaz** [comment]
Got the 4 amp power supplies and I still see the issue so it is not power starvation. 
I have found that 4 of the 5 FPP instances that were failing before are no completing successfully for some reason. That includes the one with the biggest files which drives my projector so it has lots of large sequences and video files on board. It is completing successfully with the display and I would think that would be the one that would appear to hang yet it is not and completes in 10 or 15 seconds. 
I do note that the one pi that is still failing to display (FPP-Attic) has only a small amount of relatively small compressed fseq files on it as it only drives about 500 pixels total. It does appear to complete the rsync and I see the files on the SSD. For testing I mounted the SSD and rm -rf the FPP-Attic directory structure, umount it, and then run the rsync again and see the file structure and all the files were created. So as you indicated it just does not display on the originating remote FPP.  
I am still noting that upon the failure to display the drive does not get unmounted so either I have to reboot the Master FPP or go manually umount the drive from an SSH session. Perhaps there is a better way to force the umount or provide a manual umount button in the UI?

**Comment by jaredb7** [comment]
I've pushed a small change that will help get output into the file copy backup dialogue (you should see output as it happens), 
And added headings for what area is being copied too just to give an idea of what's being copied at the time.
![image](https://github.com/FalconChristmas/fpp/assets/799998/dae832b7-5471-43d9-9b85-c191c34cb9ee)

Are you able mount the SSD, clear the existing folders for say the attic device and the projector to test those out?

I'll look into adding a button to perform a unmount if the drive is mounted, currently on say the salves, a unmount request is sent when you load the FPP backup page, and when you close the File Copy dialogue. Though If there is a copy still in progress it unmount the device, lazy unmounting sounds like it would fix all that.

**Comment by andyaz** [comment]
Tonight's experiments on this -

First I updated all FPP instances to the current release:
FPP Version: | 7.x-master-163-g4c5e84d6
Platform: | Raspberry Pi (Pi 3 Model B)
FPP OS Build: | v2023-04
OS Version: | Raspbian GNU/Linux 11 (bullseye)
Hardware Serial Number: | 00000000a2bcf78d
Kernel Version: | 6.1.21-v7+
System Boot Time: | 2023-07-04 14:51:39
fppd Uptime: | 01:58:15
Local Git Version: | 4c5e84d ChangeLog
Remote Git Version: | 4c5e84d

I manually mounted the sdb1 SSD drive on the master and manually rm- rf  all of the FPP backup directories to start fresh. 
It looks like there are some pretty large files (1gb+) on some of the remote FPPs. A good example I figured out on the FPP-Attic is it had 2 of the FPPOS images in the uploads directory which was the cause of the issue on that pi. Some of the others have large fseq files. FPP-Video had a lot of video files and large sequences for the Virtual Matrix. 
FPP instances with no large files completed and everything including the display looked perfect. 

On the ones with large files, while the display stops moving on the source FPP, I found I can monitor the progress on the master from SSH with sudo lsof /mnt/remote_filecopy/ and I do see the rsync work its way thru all of the files including the big ones. The display on the source FPP never changes when it hits the first big file. Once the rsync is completed there is no change on the display and it never prompts for a close. It also may or may not umount the remote drive in this case as well in my experiments and I can not put my finger on why that would be. In some cases I had to manually umount the SSD in others like the example below it umounted all by itself.

Display stoped here:
![image](https://github.com/FalconChristmas/fpp/assets/19541773/bcfc26af-c48e-4106-86d7-c42695fe9deb)

Monitoring the rsync progress on the master FPP using lsof:
![image](https://github.com/FalconChristmas/fpp/assets/19541773/46c6b695-5cb6-40e6-9f05-be9cb5ff3747)

rsync completes on master, it may or may not umount the remote drive. In this case it did umount:
![image](https://github.com/FalconChristmas/fpp/assets/19541773/1c52317b-3357-4f58-97d5-edbd2df0c7b6)

The display on the source FPP still looks like this when completed and never changes:
![image](https://github.com/FalconChristmas/fpp/assets/19541773/1a3c7251-fcbc-4799-a35f-0a0da6346908)

Hope this provides some insight?

**Comment by adamskiuk2002** [comment]
I don't know if this is could be the problem, but worth a check.
I'm running rsync v3.2.3 (as shipped with FPP v6.3-1-gdbe9f8b3)
On another forum I watch they were reporting problems with rsync recently and upgrading to 3.2.7 solved the issues.

**Comment by jaredb7** [comment]
> The display on the source FPP never changes when it hits the first big file. Once the rsync is completed there is no change on the display and it never prompts for a close.

Would that be one of the 1GB+ sequences? How long would that normally take to copy across the network, more than 5 minutes?
I noticed something in the logs in relation to Apache (web server) that might explain the no output or just output stopping, Apache is timing out the connection because there was no I/O happening for 300 seconds (5 minutes).

The copy script continues to run all the way through, you just never see the output due to being disconnected. I'm able to reproduce this with aggressive time out settings of 20 seconds, So bumping it way way up should resolve this issue.


>It also may or may not umount the remote drive in this case as well in my experiments and I can not put my finger on why that would be. In some cases I had to manually umount the SSD in others like the example below it umounted all by itself.

It's a bit tricky, Either the request never made it to the master or I'm thinking what I mentioned in my last comment about there being some activity still ongoing at the mount point when the unmount is requested and the unmount will just fail. If we use the lazy unmount option, as soon as that activity stops the drive is unmounted which is actually what we want in this case.

That with a button to unmount the affected device would be a good fallback.

**Comment by andyaz** [comment]
I would bump it up to say 15 minutes of no activity to time out. If, and I know that is probably a big ask to pull off, but if there was a method to display the fact that a timeout occurred to the display would be ideal so there would be no doubt when it happens vs staring at a never changing progress screen. Did you figure out if it is PHP or apache timing out? 

Something I found on the web that might be related:
[https://stackoverflow.com/questions/49612459/can-i-increase-apaches-timeout-directive-for-specific-php-script](url)
**<snip>
<Question>**
I have a PHP script that will be executed by requests from the application admins. It does lots of stuff and takes at least 20 minutes(depending on the database size),  the Apache TimeOut directive is set to 300(5 minutes) which closes the connection and returns 500 status code after my PHP script is finished if it takes longer time to execute.
Setting the PHP ini max_execution_time for long time for this script is useless and I don't want to change the entire Apache TimeOut directive just for those couple of scripts, because if I did, any request will be able to take very long time which makes a scope for DDOS vulnerabilities, is this right?
Is there any way to allow this script only to run longer at the Apache level ?

**<Answer>** 
Have you tried PHP's set_time_limit() method?
https://www.php.net/manual/en/function.set-time-limit.php
In addition to setting the initial execution time, the manual says that calling it resets the time expended to zero when called, and starts the counter again to the limit provided.
So if you want to be sure, you could just call set_time_limit(0) (0 == no limit) regularly throughout your script to make sure that you don't ever hit a limit (even though you're setting an infinite limit by passing in 0).
**</snip>**

Wonder if this would help in this case of the large files?

**Comment by jaredb7** [comment]
> Did you figure out if it is PHP or apache timing out?

Apache time-out, well probably both that and the apache fcgi proxy module timing out ( have both set to 20 seconds for testing, the time-out error is to do with the fcgi proxy), which appears to handle connections to a process executing the requested PHP page.

I have tried set_time_limit(20) to cause the script to time-out early since I thought that was one of the issues but it didn't have the desired effect. 
I was thinking of setting this to 1 hr anyway to give the whole script headroom to complete.
I've put in ignore_user_abort(true); a few commits ago so the script continue to run if the browser is disconnected. I've had success using both on long running standalone scripts

> I know that is probably a big ask to pull off, but if there was a method to display the fact that a timeout occurred to the display would be ideal so there would be no doubt when it happens vs staring at a never changing progress screen

Easy done, Initially I just had a message explaining what might have happened... But after some thinking, I come up with a bit better solution by using `tee` and logging the same output to a log.. then switch over to reading that after the connection gets terminated by Apache. Since we're requesting the log every second this method won't time out and we can still keep an eye on the process we started (it continues to run) but timed out on
![image](https://github.com/FalconChristmas/fpp/assets/799998/625d1463-2995-426c-8571-a5f28d15469b)

I'll flesh this out more and implement a way to force unmounting over the next few days

**Comment by jaredb7** [comment]
I've implemented the ability to unmount the USB drive under the Storage tab of FPP Settings, basic drive info is shown and a button to perform the unmount is available.
lsof output is also available if there are open files and is just to give some feedback to the user as to what might be using it.
The Open files section is hidden if lsof output is empty.
![image](https://github.com/FalconChristmas/fpp/assets/799998/db980a07-e059-45c0-893d-9fda5e9c13ce)

The File Copy Backup page now displays a message if your chosen remote storage/USB device is not available in the list of available USB devices. A link to the chosen remote host FPP Setting is available so you can jump across and investigate. 
As you've experience most times this is because it's still mounted, but could also be if the device is really not available so the message tries to address both
![image](https://github.com/FalconChristmas/fpp/assets/799998/bd33e745-ef71-4297-85db-32fd45947832)
 
I also created a alternative method to get the output/process of the copy process in the event Apache terminates the connection, the dialogue will switch over to this automatically if this happens.
 
Increased the Apache ProxyTimeout setting that seemed to terminate connections to the PHP process from 5 to 20 mins, the script appears to actually run until completion anyway just the connection to the process was dropped.
We might need to also increase the Timeout value but the errors were in relation to the PHP FPM proxy so we'll see how this goes first.

If you could put this through it's paces that would be great.

**Comment by andyaz** [comment]
Nicely done. First test was 3.4 GB and it looks like it worked. No updates on the display once the big transfer started but it completed and the display finished out and showed the final stats and unmount action.

I did see this: !!! A connection error or timeout has occurred, unable to continue tracking file copy progress... The copy is still running in the background and will complete in due course. Attempting to tail it's fallback log file... !!! 

I also noted some other strange behavior where the display looked like it dumped the config settings to the screen... I need to do more testing to try and capture what I see. Is there a log file (like that fall back log) I can look at to see what it was sending to the display start to finish?

![image](https://github.com/FalconChristmas/fpp/assets/19541773/b1f494a1-55ce-4e5b-9029-da817fec4106)

**Comment by andyaz** [comment]
I did some digging in your code to find that backup file so I am including a screen grab of it as well in here. Here are a couple screen grabs from the display... This shows the appearance of a config dump in the display but not in the backup log file you create.

Start of backup display
![image](https://github.com/FalconChristmas/fpp/assets/19541773/90294fd9-00b0-44ba-95e0-5b85be9aee9e)

End of the "config dump" and start of transfer
![image](https://github.com/FalconChristmas/fpp/assets/19541773/188a38fc-ca5d-474f-a5de-34cd13a65de8)

Same a few line further down as transfer begins. Again transferring the FPPOS file to have something huge to test with
![image](https://github.com/FalconChristmas/fpp/assets/19541773/f294f562-9686-49a2-8746-5eae0384a897)

The fpp_backup_filecopy.log at the same place in time
![image](https://github.com/FalconChristmas/fpp/assets/19541773/5ea5d1cb-e227-4cfe-8860-752b06652bd7)

 I will try to get some of the other strangeness I saw displayed as well as the transfer progresses...

**Comment by andyaz** [comment]
A few minutes later this is the display with no config dump above it
![image](https://github.com/FalconChristmas/fpp/assets/19541773/5780043d-98d9-407a-86ba-747f18367461)

**Comment by andyaz** [comment]
A few minutes later the display changes to this as expected
![image](https://github.com/FalconChristmas/fpp/assets/19541773/2b4a5541-7789-45c1-b0c2-03b1f524ac5d)

**Comment by andyaz** [comment]
A few minutes later we get this, I believe as expected
![image](https://github.com/FalconChristmas/fpp/assets/19541773/d41851ed-7da5-4026-8022-0dcfd279e3ff)

**Comment by andyaz** [comment]
Ahh, note the close tab but the transfer is still running in the background as you can see by the lsof on the master. I will wait it out and see what it does from here without hitting the close.

![image](https://github.com/FalconChristmas/fpp/assets/19541773/6909bf8f-800d-4c22-99c6-3a0d47f45251)

**Comment by andyaz** [comment]
Transfer completed, no change to the display from last snip with the 900 second timeout. Here is the output of the fpp_backup log before it closed
![image](https://github.com/FalconChristmas/fpp/assets/19541773/75258734-1abb-4c2e-ba4a-ee234f16b4e0)

**Comment by andyaz** [comment]
I see you pushed more changes in GitHub... I updated and below is the version detail for the tested FPP for your reference -
FPP Version:	                  7.x-master-189-g19714c45
Platform:	                          Raspberry Pi (Pi 3 Model B+)
FPP OS Build:	                  v2023-06
OS Version:	                  Raspbian GNU/Linux 11 (bullseye)
Kernel Version:	         6.1.21-v7+
System Boot Time:	         2023-07-04 14:51:41
fppd Uptime:	                 56:39
Local Git Version:	        19714c4 
Remote Git Version:	19714c4

I tested on my FPP-VIDEO (runs the virtual matrix) which has more large files than others. Mounted the SSD and deleted the backup directory for it first, and then ran the copy backup. This ran nicely with no noted weirdness in the display and it tracked all the way thru the 3.44G of transfers including the 1.5G of FPPOS and all of the large and small files in this FPP instance. I did note the screen refresh at one point during the FPPOS transfer where it added the notice for tracking the backup log file and filled all the details back in to the current point in the process. It did continue to update the display for each remaining transfer nicely after that refresh with no noted hiccups and the whole thing completed successfully. 

Nice job!

![image](https://github.com/FalconChristmas/fpp/assets/19541773/7af586a9-eb42-4e20-957e-88fc7f5d7d6e)


![image](https://github.com/FalconChristmas/fpp/assets/19541773/5f5a92e6-b048-47d6-9ed7-4e10034ef11c)

**Comment by jaredb7** [comment]
Awesome! :). Also thanks for the last round of screenshots, especially the last one of the rsync log. It was good to see really how it was taking to move what I presume is the FFPOS file at 32 minutes.

I pushed the changes last night but it was getting very late so just left it at that.

In the changes I made I decided to rename the fpp_backup_filecopy.log to fpp_backup_filecopy_last.log at the end just so we have a copy if we ever need it.
